pkg/utils/controller/controller_utils.go

```diff
@@ -61,3 +72,213 @@ func CreateHeadlessServiceIfNotExists(ctx context.Context, k8sClient client.Clie
+	selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{MatchLabels: map[string]string{
+		leaderworkerset.TemplateRevisionHashKey: templateHash,
+	}})
+	if err != nil {
+		return nil, err
+	}
+	revisions, err := controllerHistory.ListControllerRevisions(lws, selector)
+	if err != nil {
+		log.Error(err, "Listing all controller revisions")
+		return nil, err
+	}
+
+	if len(revisions) == 0 {
+		return nil, fmt.Errorf("could not find LWS revision based on %s", templateHash)
+	}
+
+	if len(revisions) > 1 {
+		// Since we only create a controllerRevision when the template hash changes, only one should match
+		return nil, fmt.Errorf("found more than one revision matching templateHash %s", templateHash)
+	}
+
+	return revisions[0], nil
+}
+
+func ExistingControllerRevisions(ctx context.Context, k8sClient client.Client, lws *leaderworkerset.LeaderWorkerSet) (bool, error) {
+	log := ctrl.LoggerFrom(ctx).WithValues("leaderworkerset", klog.KObj(lws))
+	ctx = ctrl.LoggerInto(ctx, log)
+	controllerHistory := history.NewHistory(ctx, k8sClient)
+	selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{})
+	if err != nil {
+		return false, err
+	}
+	revisions, err := controllerHistory.ListControllerRevisions(lws, selector)
+	if err != nil {
+		return false, err
+	}
+	return len(revisions) > 0, nil
+}
+
+// getPatch returns a strategic merge patch that can be applied to restore a LeaderWorkerSet to a
+// previous version. If the returned error is nil the patch is valid. The current state that we save is the
+// leaderWorkerTemplate. We can modify this later to encompass more state (or less) and remain compatible with previously
+// recorded patches.
+
+func getPatch(lws *leaderworkerset.LeaderWorkerSet) ([]byte, error) {
+	str := &bytes.Buffer{}
+	clone := lws.DeepCopy()
+	err := unstructured.UnstructuredJSONScheme.Encode(clone, str)
+	if err != nil {
+		return nil, err
+	}
+	var raw map[string]interface{}
+	err = json.Unmarshal(str.Bytes(), &raw)
+	objCopy := make(map[string]interface{})
+	specCopy := make(map[string]interface{})
+	spec := raw["spec"].(map[string]interface{})
+	specCopy = spec
+	specCopy["$patch"] = "replace"
+	objCopy["spec"] = spec
+	if err != nil {
+		return nil, err
+	}
+	patch, err := json.Marshal(objCopy)
+	return patch, err
+}
+
+func CreateLeaderWorkerSetRevision(
+	ctx context.Context,
+	k8sClient client.Client,
+	lws *leaderworkerset.LeaderWorkerSet,
+	templateHash string) error {
+	log := ctrl.LoggerFrom(ctx).WithValues("leaderworkerset", klog.KObj(lws))
+	ctx = ctrl.LoggerInto(ctx, log)
+	controllerHistory := history.NewHistory(ctx, k8sClient)
+	selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{})
+	if err != nil {
+		return err
+	}
+	revisions, err := controllerHistory.ListControllerRevisions(lws, selector)
+	if err != nil {
+		log.Error(err, "Listing all controller revisions")
+		return err
+	}
+	revisionCount := len(revisions)
+	history.SortControllerRevisions(revisions)
```

Comment:
No need to sort, I would simplify and remove the sort functions, and just iterate here and find the max revision in the list. If the list is 0, use revision=1

Reply:
Changed logic of `NextRevision`